File: filesystem/v2_example.py
"""Manage v2 repositories from Python through their HTTP API.

The Letta Agent SDK is currently TypeScript-only. This Python example covers
repository creation, text-file writes, listing/version history, and optional
attachment to an existing agent. Use v2_example.ts when you also need to create
a managed Agent SDK session and ask the agent questions about the repository.
"""

import os
import time
from pathlib import Path
from typing import Any, Dict

import httpx

api_key = os.environ.get("LETTA_API_KEY")
if not api_key:
    raise RuntimeError("LETTA_API_KEY is required")

paper_text = (
    Path(os.environ.get("PAPER_TEXT", "memgpt.txt"))
    .resolve()
    .read_text(encoding="utf-8", errors="replace")
)
agent_id = os.environ.get("TARGET_AGENT_ID")
suffix = format(int(time.time() * 1000), "x")

repository_id = None
attached = False

with httpx.Client(
    base_url="https://api.letta.com",
    headers={"Authorization": f"Bearer {api_key}"},
    timeout=120,
) as client:

    def api(method: str, path: str, **kwargs: Any) -> Dict[str, Any]:
        response = client.request(method, path, **kwargs)
        response.raise_for_status()
        return response.json()

    try:
        repository = api(
            "POST",
            "/v1/repositories",
            json={"name": f"memgpt-paper-v2-python-{suffix}"},
        )
        repository_id = repository["id"]
        print(f"repository: {repository_id}")

        mutation = api(
            "POST",
            f"/v1/repositories/{repository_id}/files",
            json={
                "path": "papers/memgpt.txt",
                "content": paper_text,
            },
        )
        print(f"file commit: {mutation['commit_sha']}")

        files = api(
            "GET",
            f"/v1/repositories/{repository_id}/files",
            params={"path_prefix": "papers/"},
        )
        print("files:", ", ".join(item["path"] for item in files["files"]))

        versions = api(
            "GET",
            f"/v1/repositories/{repository_id}/versions",
            params={"path": "papers/memgpt.txt", "limit": 20},
        )
        commits = versions.get("commits", versions.get("versions", versions))
        print(f"versions: {len(commits)}")

        if agent_id:
            api(
                "POST",
                f"/v1/agents/{agent_id}/repositories",
                json={"repository_id": repository_id},
            )

            deadline = time.monotonic() + 10
            while time.monotonic() < deadline:
                linked = api("GET", f"/v1/agents/{agent_id}/repositories")
                if any(repo["id"] == repository_id for repo in linked["repositories"]):
                    attached = True
                    break
                time.sleep(0.25)

            if not attached:
                raise TimeoutError("Repository attachment did not become visible")
            print(f"attached to agent: {agent_id}")
        else:
            print(
                "Set TARGET_AGENT_ID to demonstrate persistent attachment. "
                "Use v2_example.ts for the full session-scoped document Q&A flow."
            )
    finally:
        if os.environ.get("KEEP_RESOURCES") != "1":
            if attached and agent_id and repository_id:
                response = client.delete(
                    f"/v1/agents/{agent_id}/repositories/{repository_id}"
                )
                if response.status_code not in (200, 204, 404):
                    response.raise_for_status()

            if repository_id:
                response = client.delete(f"/v1/repositories/{repository_id}")
                if response.status_code not in (200, 204, 404):
                    response.raise_for_status()

File: filesystem/v2_example.ts
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";
import { LettaAgentClient } from "@letta-ai/letta-agent-sdk";

const apiKey = process.env.LETTA_API_KEY;
if (!apiKey) throw new Error("LETTA_API_KEY is required");

// Repository files are text content, so extract text from the source PDF first.
const paperText = await readFile(
  resolve(process.env.PAPER_TEXT ?? "memgpt.txt"),
  "utf8",
);
const suffix = Date.now().toString(36);
const client = new LettaAgentClient({
  backend: "cloud",
  apiKey,
  apiBaseUrl: "https://api.letta.com",
  sandbox: { terminateOnClose: true },
});

let agentId: string | undefined;
let repositoryId: string | undefined;
let session: ReturnType<typeof client.createSession> | undefined;

async function deleteAgent(id: string) {
  const response = await fetch(
    `https://api.letta.com/v1/agents/${encodeURIComponent(id)}`,
    {
      method: "DELETE",
      headers: { Authorization: `Bearer ${apiKey}` },
    },
  );

  if (!response.ok && response.status !== 404) {
    throw new Error(`Agent cleanup failed with HTTP ${response.status}`);
  }
}

try {
  const repository = await client.repositories.create({
    name: `memgpt-paper-v2-${suffix}`,
  });
  repositoryId = repository.id;
  console.log(`repository: ${repository.id}`);

  const mutation = await client.repositories.files.create(repository.id, {
    path: "papers/memgpt.txt",
    content: paperText,
  });
  console.log(`file commit: ${mutation.commitSha}`);

  const files = await client.repositories.files.list(repository.id, {
    pathPrefix: "papers/",
  });
  console.log(`files: ${files.files.map((file) => file.path).join(", ")}`);

  agentId = await client.createAgent({
    name: `memgpt-repositories-example-${suffix}`,
    model: "openai/gpt-5.2",
    persona:
      "You are a careful research assistant. Answer document questions from attached repositories and quote the supporting text.",
    memfs: false,
  });
  console.log(`agent: ${agentId}`);

  // The SDK attaches this repository before startup and detaches it on close.
  session = client.createSession(agentId, {
    permissionMode: "unrestricted",
    resources: [{ type: "repository", repositoryId: repository.id }],
  });

  const question = `Use the attached repository ${repository.name}, specifically papers/memgpt.txt, to answer these questions. Cite short supporting excerpts from the paper for each answer.

1. What operating-system mechanism inspired MemGPT's virtual context management, and how does the paper map the LLM's main and external context to that mechanism?
2. Which two domains were used to evaluate MemGPT, and what new synthetic retrieval task did the paper introduce?`;

  await session.send(question);

  const assistantParts: string[] = [];
  const toolCalls = new Map<string, string>();

  for await (const message of session.stream()) {
    if (message.type === "tool_call") {
      toolCalls.set(message.toolCallId, message.toolName);
    } else if (message.type === "assistant") {
      assistantParts.push(message.content);
    } else if (message.type === "error") {
      throw new Error(message.message);
    } else if (message.type === "result" && !message.success) {
      throw new Error(
        message.errorDetail ?? message.error ?? "Agent turn failed",
      );
    }
  }

  console.log(`tools used: ${[...new Set(toolCalls.values())].join(", ")}`);
  console.log(`\nANSWER\n${assistantParts.join("")}`);
} finally {
  session?.close();

  if (process.env.KEEP_RESOURCES !== "1") {
    // Session detach and managed-sandbox cleanup are asynchronous and best effort.
    await new Promise((resolveDelay) => setTimeout(resolveDelay, 2_000));
    if (repositoryId) {
      await client.repositories.delete(repositoryId).catch(() => undefined);
    }
    if (agentId) await deleteAgent(agentId).catch(() => undefined);
  }
}
